### DISCUSSION TITLE
추상화 계층의 테스트 코드를 작성할 때

### DISCUSSION THREAD

**Discussion by kth990303** [woowacourse-teams/2022-nae-pyeon#21] [BE 테스트 설계 이슈]
## Repository 테스트를 작성할 때

현재 구조는 Repository와 이를 구현하고 있는 JpaRepository가 있습니다.
이것에 대한 테스트를 작성할 때, 어떤 테스트를 작성해야 된다고 생각하시나요?

1. Repository 인터페이스 테스트 + JpaRepository 구현체 테스트
2. Repository 인터페이스만 테스트 (`@SpringBootTest`를 활용)
3. JpaRepository 구현체만 테스트 

저랑 알렉스는 1번 또는 2번이 좋다고 생각하였고, 현재 2번 방법으로 작성했습니다. 그리고 제로랑 승팡은 3번 방법으로 테스트를 한 것 같아요.

저랑 알렉스가 생각한 1번 또는 2번이 옳다고 생각한 이유는 아래와 같습니다.
- 추상화된 인터페이스를 테스트하여 통과한다는 것은, 그 구현체 테스트가 충분히 이루어지는 경우라 생각.
- **우리가 직접 이용하는 것은 추상화된 고수준 인터페이스이므로**, 이걸 테스트해야 된다고 생각. (3번 방법을 사용하지 않는 이유)
  - 우리는 JPA에 종속되지 않은 repository 인터페이스를 만들었으므로 이 repository 인터페이스를 테스트하는 것이 맞다고 생각
- 추상화된 인터페이스를 테스트한다면, 우리가 사용하지 않는 구현체들은 테스트되지 않는 것이 아닌가? -> 구현체를 갈아끼우는 경우, 우리가 직접적으로 사용하는 구현체를 추상화한 인터페이스를 테스트하므로 우리가 사용하는 구현체들이 테스트되므로 신뢰성에 문제가 없음.

제로랑 승팡이 `JpaMemberRepositoryTest`만 작성하셨던데, 
여기에 추가로 `MemberRepositoryTest`를 작성하거나, 아니면 아예 `MemberRepositoryTest`만 있는 것이 좋다고 생각합니다.
제로랑 승팡 생각은 어떤지 궁금합니다!

**Comment by yxxnghwan** [comment]
저 고민을 하게 된 구체적인 상황을 설명드리자면,

1. Jpa를 사용한 구현체를 테스트할 때 필드로 가지는 협력 엔티티를 extracting 해서 검증할 경우 Jpa의 영속성 컨테스트에서 관리되는 엔티티 동일성 보장 기능 덕에 동일한 객체로 인식해서 테스트가 통과했습니다.
2. 하지만 만약 Jpa의 동일성 보장 기능이 없었다면 해당 테스트가 통과할 수 없었을 거라고 생각해요. `JpaRollingpaperRepository`는 Jpa의 기술에 의존적인 구현체이므로 이 구현체에 대한 테스트로는 충분하지만, 우리가 주입받아 사용하는 `RollingpaperRepository`라는 인터페이스의 테스트는 아니라는 생각이 들었습니다.
3. 그래서 구현 기술에 의존적이지 않은 테스트 코드가 필요하다고 생각해서 구현체와 인터페이스 둘 다 테스트 할 건지, 인터페이스만 테스트할 건지 고민하게 되었어요!

백엔드끼리 한 번 같이 고민해봐야하는 문제인 것 같아서 디스커션 등록합니다!💪

**Comment by asebn1** [comment]
공감합니다
SpringBootTest + MemberRepositoryTest 로 적용해보도록 하겠습니다!
